File: app/src/main/java/locus/api/mapper/GeocacheLogConverter.java
package locus.api.mapper;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import com.arcao.geocaching.api.data.GeocacheLog;
import com.arcao.geocaching.api.data.ImageData;
import com.arcao.geocaching.api.data.User;
import com.arcao.geocaching.api.data.type.GeocacheLogType;
import java.util.Collection;
import java.util.Collections;
import java.util.Comparator;
import locus.api.objects.extra.Waypoint;
import locus.api.objects.geocaching.GeocachingLog;
import org.apache.commons.collections4.CollectionUtils;
import org.jetbrains.annotations.NotNull;

import static locus.api.mapper.Util.safeDateLong;

final class GeocacheLogConverter {
  private final ImageDataConverter imageDataConverter;

  GeocacheLogConverter(ImageDataConverter imageDataConverter) {
    this.imageDataConverter = imageDataConverter;
  }

  void addGeocacheLogs(@NonNull Waypoint waypoint, @Nullable Collection<GeocacheLog> logs) {
    if (waypoint.gcData == null || CollectionUtils.isEmpty(logs))
      return;

    for (GeocacheLog log : logs) {
      CollectionUtils.addIgnoreNull(waypoint.gcData.logs, createLocusGeocachingLog(log));
    }

    sortLocusGeocachingLogsByDate(waypoint);
  }

  @Nullable
  private GeocachingLog createLocusGeocachingLog(@Nullable GeocacheLog log) {
    if (log == null)
      return null;

    GeocachingLog l = new GeocachingLog();
    l.setId(log.id());
    l.setDate(safeDateLong(log.visited()));

    User author = log.author();
    if (author != null) {
      l.setFinder(author.userName());
      l.setFindersFound(author.findCount());
      l.setFindersId(author.id());
    }

    l.setLogText(log.text());
    l.setType(createLocusCacheLogType(log.logType()));

    for (ImageData image: log.images()) {
      l.addImage(imageDataConverter.createLocusGeocachingImage(image));
    }

    if (log.updatedCoordinates() != null) {
      l.setCooLat(log.updatedCoordinates().latitude());
      l.setCooLon(log.updatedCoordinates().longitude());
    }

    return l;
  }


  private int createLocusCacheLogType(@Nullable GeocacheLogType logType) {
    if (logType == null)
      return  GeocachingLog.CACHE_LOG_TYPE_UNKNOWN;

    switch (logType) {
      case Announcement:
        return GeocachingLog.CACHE_LOG_TYPE_ANNOUNCEMENT;
      case Attended:
        return GeocachingLog.CACHE_LOG_TYPE_ATTENDED;
      case DidntFindIt:
        return GeocachingLog.CACHE_LOG_TYPE_NOT_FOUND;
      case EnableListing:
        return GeocachingLog.CACHE_LOG_TYPE_ENABLE_LISTING;
      case FoundIt:
        return GeocachingLog.CACHE_LOG_TYPE_FOUND;
      case NeedsArchived:
        return GeocachingLog.CACHE_LOG_TYPE_NEEDS_ARCHIVED;
      case NeedsMaintenance:
        return GeocachingLog.CACHE_LOG_TYPE_NEEDS_MAINTENANCE;
      case OwnerMaintenance:
        return GeocachingLog.CACHE_LOG_TYPE_OWNER_MAINTENANCE;
      case PostReviewerNote:
        return GeocachingLog.CACHE_LOG_TYPE_POST_REVIEWER_NOTE;
      case PublishListing:
        return GeocachingLog.CACHE_LOG_TYPE_PUBLISH_LISTING;
      case RetractListing:
        return GeocachingLog.CACHE_LOG_TYPE_RETRACT_LISTING;
      case TemporarilyDisableListing:
        return GeocachingLog.CACHE_LOG_TYPE_TEMPORARILY_DISABLE_LISTING;
      case Unknown:
        return GeocachingLog.CACHE_LOG_TYPE_UNKNOWN;
      case UpdateCoordinates:
        return GeocachingLog.CACHE_LOG_TYPE_UPDATE_COORDINATES;
      case WebcamPhotoTaken:
        return GeocachingLog.CACHE_LOG_TYPE_WEBCAM_PHOTO_TAKEN;
      case WillAttend:
        return GeocachingLog.CACHE_LOG_TYPE_WILL_ATTEND;
      case WriteNote:
        return GeocachingLog.CACHE_LOG_TYPE_WRITE_NOTE;
      case Archive:
        return GeocachingLog.CACHE_LOG_TYPE_ARCHIVE;
      case Unarchive:
        return GeocachingLog.CACHE_LOG_TYPE_UNARCHIVE;
      default:
        return GeocachingLog.CACHE_LOG_TYPE_UNKNOWN;
    }
  }

  private void sortLocusGeocachingLogsByDate(@NotNull Waypoint waypoint) {
    if (waypoint.gcData == null || CollectionUtils.isEmpty(waypoint.gcData.logs))
      return;

    Collections.sort(waypoint.gcData.logs, new Comparator<GeocachingLog>() {
      @Override
      public int compare(GeocachingLog lhs, GeocachingLog rhs) {
        return lhs.getDate() > rhs.getDate() ? 1 : lhs.getDate() == rhs.getDate() ? -1 : 0;
      }
    });
  }

}

File: app/src/main/java/locus/api/mapper/GeocacheLogConverter.kt
package locus.api.mapper

import com.arcao.geocaching4locus.data.api.model.GeocacheLog
import com.arcao.geocaching4locus.data.api.model.GeocacheLogType
import locus.api.objects.geoData.Point
import locus.api.objects.geocaching.GeocachingLog
import locus.api.utils.addIgnoreNull

class GeocacheLogConverter(
    private val imageDataConverter: ImageDataConverter
) {

    fun addGeocacheLogs(point: Point, logs: Collection<GeocacheLog>) {
        if (point.gcData?.logs == null || logs.isEmpty())
            return

        for (log in logs) {
            point.gcData?.logs?.addIgnoreNull(createLocusGeocachingLog(log))
        }

        sortLocusGeocachingLogsByDate(point)
    }

    fun createLocusGeocachingLogs(logs: Collection<GeocacheLog>): Collection<GeocachingLog> {
        return logs.map(this::createLocusGeocachingLog)
    }

    private fun createLocusGeocachingLog(log: GeocacheLog): GeocachingLog {
        return GeocachingLog().apply {
            id = log.id
            date = log.loggedDateInstant?.toEpochMilli() ?: 0
            logText = log.text.orEmpty()
            type = log.geocacheLogType.toLocusMapLogType()

            val author = log.owner
            if (author != null) {
                finder = author.username.orEmpty()
                findersFound = author.findCount
                findersId = author.id
            }

            for (image in log.images ?: emptyList()) {
                imageDataConverter.createLocusGeocachingImage(image)?.let(this::addImage)
            }
            log.updatedCoordinates?.let { coordinates ->
                cooLat = coordinates.latitude
                cooLon = coordinates.longitude
            }
        }
    }

    private fun GeocacheLogType?.toLocusMapLogType(): Int {
        return when (this?.id ?: GeocacheLogType.WRITE_NOTE) {
            GeocacheLogType.EVENT_ANNOUNCEMENT -> GeocachingLog.CACHE_LOG_TYPE_ANNOUNCEMENT
            GeocacheLogType.ATTENDED -> GeocachingLog.CACHE_LOG_TYPE_ATTENDED
            GeocacheLogType.DNF_IT -> GeocachingLog.CACHE_LOG_TYPE_NOT_FOUND
            GeocacheLogType.ENABLE_LISTING -> GeocachingLog.CACHE_LOG_TYPE_ENABLE_LISTING
            GeocacheLogType.FOUND_IT -> GeocachingLog.CACHE_LOG_TYPE_FOUND
            GeocacheLogType.NEEDS_ARCHIVING -> GeocachingLog.CACHE_LOG_TYPE_NEEDS_ARCHIVED
            GeocacheLogType.NEEDS_MAINTENANCE -> GeocachingLog.CACHE_LOG_TYPE_NEEDS_MAINTENANCE
            GeocacheLogType.OWNER_MAINTENANCE -> GeocachingLog.CACHE_LOG_TYPE_OWNER_MAINTENANCE
            GeocacheLogType.POST_REVIEWER_NOTE -> GeocachingLog.CACHE_LOG_TYPE_POST_REVIEWER_NOTE
            GeocacheLogType.PUBLISH_LISTING -> GeocachingLog.CACHE_LOG_TYPE_PUBLISH_LISTING
            GeocacheLogType.RETRACT_LISTING -> GeocachingLog.CACHE_LOG_TYPE_RETRACT_LISTING
            GeocacheLogType.TEMPORARILY_DISABLE_LISTING -> GeocachingLog.CACHE_LOG_TYPE_TEMPORARILY_DISABLE_LISTING
            GeocacheLogType.UPDATE_COORDINATES -> GeocachingLog.CACHE_LOG_TYPE_UPDATE_COORDINATES
            GeocacheLogType.WEBCAM_PHOTO_TAKEN -> GeocachingLog.CACHE_LOG_TYPE_WEBCAM_PHOTO_TAKEN
            GeocacheLogType.WILL_ATTEND -> GeocachingLog.CACHE_LOG_TYPE_WILL_ATTEND
            GeocacheLogType.WRITE_NOTE -> GeocachingLog.CACHE_LOG_TYPE_WRITE_NOTE
            GeocacheLogType.ARCHIVE -> GeocachingLog.CACHE_LOG_TYPE_ARCHIVE
            GeocacheLogType.UNARCHIVE -> GeocachingLog.CACHE_LOG_TYPE_UNARCHIVE
            GeocacheLogType.SUBMIT_FOR_REVIEW -> GeocachingLog.CACHE_LOG_TYPE_WRITE_NOTE
            else -> GeocachingLog.CACHE_LOG_TYPE_WRITE_NOTE
        }
    }

    private fun sortLocusGeocachingLogsByDate(waypoint: Point) {
        if (waypoint.gcData?.logs?.isEmpty() != false)
            return

        waypoint.gcData?.logs?.sortWith { lhs, rhs -> lhs.date.compareTo(rhs.date) }
    }
}
